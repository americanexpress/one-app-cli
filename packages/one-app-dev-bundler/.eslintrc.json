{
  "extends": "amex",
  "rules": {
    "no-console": 0,
    "import/extensions": 0 // ECMAscript modules explicitly require all extensions for relative imports
  }
}
