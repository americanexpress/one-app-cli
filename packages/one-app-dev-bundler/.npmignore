# test
__tests__
__mocks__

# dev
.eslintrc.json

# npm pack
*.tgz
