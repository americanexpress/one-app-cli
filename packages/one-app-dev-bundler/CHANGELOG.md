# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.5.1](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.5.0...@americanexpress/one-app-dev-bundler@1.5.1) (2023-09-15)


### Bug Fixes

* **externals:** sanatize external name ([#563](https://github.com/americanexpress/one-app-cli/issues/563)) ([4be6f3b](https://github.com/americanexpress/one-app-cli/commit/4be6f3be3ae1838cf09d0438b051382a9c206747))





# [1.5.0](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.4.2...@americanexpress/one-app-dev-bundler@1.5.0) (2023-09-14)


### Bug Fixes

* **dev-bundler:** use semver ranges for all deps ([#550](https://github.com/americanexpress/one-app-cli/issues/550)) ([991ea3b](https://github.com/americanexpress/one-app-cli/commit/991ea3be4bea76f9a80200255a2f2e22a1180929))


### Features

* **devBundler:** aggregate styles into dependencies and local ([#557](https://github.com/americanexpress/one-app-cli/issues/557)) ([0eeee49](https://github.com/americanexpress/one-app-cli/commit/0eeee499ee9fda2eac6b2663c754c6840053d05a))
* external fallbacks ([#536](https://github.com/americanexpress/one-app-cli/issues/536)) ([523898d](https://github.com/americanexpress/one-app-cli/commit/523898deb9a1a4bcce6ba43915c852b02b7bb3a5))





## [1.4.2](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.4.1...@americanexpress/one-app-dev-bundler@1.4.2) (2023-06-13)


### Bug Fixes

* **one-app-dev-bundler:** avoid default export of sass ([0d9b5b4](https://github.com/americanexpress/one-app-cli/commit/0d9b5b422a9e18cc6391fa119abe41b7d384c154))





## [1.4.1](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.4.0...@americanexpress/one-app-dev-bundler@1.4.1) (2023-05-15)


### Bug Fixes

* **dev-bundler:** improve error when NODE_ENV not development ([#528](https://github.com/americanexpress/one-app-cli/issues/528)) ([e1babee](https://github.com/americanexpress/one-app-cli/commit/e1babeed1ddc3f0e41f49850e9eb74bf494d0aa9))





# [1.4.0](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.3.1...@americanexpress/one-app-dev-bundler@1.4.0) (2023-05-03)


### Bug Fixes

* **dev-bundler:** skip plugins when no results metadata ([#513](https://github.com/americanexpress/one-app-cli/issues/513)) ([76b5566](https://github.com/americanexpress/one-app-cli/commit/76b5566b1d74a5686c31fbbb24c496e6b8af59db))
* **devBundler:** upgrade to cssnano@5 ([#518](https://github.com/americanexpress/one-app-cli/issues/518)) ([5abc78b](https://github.com/americanexpress/one-app-cli/commit/5abc78b096c89629c4a8ba6418282d498b09c28f))
* **one-app-dev-bundler:** index loader filter for Windows paths ([#507](https://github.com/americanexpress/one-app-cli/issues/507)) ([a6b66e9](https://github.com/americanexpress/one-app-cli/commit/a6b66e99d7b9e77b76714b772c9d7f116191bcc0))
* **one-app-dev-bundler:** match a period, rather than wildcard ([#508](https://github.com/americanexpress/one-app-cli/issues/508)) ([490f74b](https://github.com/americanexpress/one-app-cli/commit/490f74b43aca5bf1cb1bd0cb88deab2af60c11b7))


### Features

* **typeScript:** allow bundlers to consume ts files ([#520](https://github.com/americanexpress/one-app-cli/issues/520)) ([77cb796](https://github.com/americanexpress/one-app-cli/commit/77cb796ed028cae25d3fcc38171d84594d15e3d3))





## [1.3.2](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.3.1...@americanexpress/one-app-dev-bundler@1.3.2) (2023-04-07)


### Bug Fixes

* **dev-bundler:** skip plugins when no results metadata ([#513](https://github.com/americanexpress/one-app-cli/issues/513)) ([76b5566](https://github.com/americanexpress/one-app-cli/commit/76b5566b1d74a5686c31fbbb24c496e6b8af59db))
* **one-app-dev-bundler:** index loader filter for Windows paths ([#507](https://github.com/americanexpress/one-app-cli/issues/507)) ([a6b66e9](https://github.com/americanexpress/one-app-cli/commit/a6b66e99d7b9e77b76714b772c9d7f116191bcc0))
* **one-app-dev-bundler:** match a period, rather than wildcard ([#508](https://github.com/americanexpress/one-app-cli/issues/508)) ([490f74b](https://github.com/americanexpress/one-app-cli/commit/490f74b43aca5bf1cb1bd0cb88deab2af60c11b7))





## [1.3.1](https://github.com/americanexpress/one-app-cli/compare/@americanexpress/one-app-dev-bundler@1.3.0...@americanexpress/one-app-dev-bundler@1.3.1) (2023-03-16)


### Bug Fixes

* **devBundler:** support named exports from style files ([#502](https://github.com/americanexpress/one-app-cli/issues/502)) ([0db64a1](https://github.com/americanexpress/one-app-cli/commit/0db64a152d762bc1489e25f951622f5667718478))





# 1.3.0 (2023-03-08)


### Bug Fixes

* **devBundler:** add node polyfill ([#497](https://github.com/americanexpress/one-app-cli/issues/497)) ([56bb14a](https://github.com/americanexpress/one-app-cli/commit/56bb14abb09faae2efe53e009285a1149c56680d))
* **devBundler:** add node polyfill ([#499](https://github.com/americanexpress/one-app-cli/issues/499)) ([dd729f3](https://github.com/americanexpress/one-app-cli/commit/dd729f38adb0a1450564333b2745eee8e3070625))
* **indexloader:** regex was failing on windows machines ([#494](https://github.com/americanexpress/one-app-cli/issues/494)) ([aadd37f](https://github.com/americanexpress/one-app-cli/commit/aadd37faea92fce9b6c691c061441e1624f40cb1))


### Features

* **bundler:** enable dev bundler through flag ([1cc171b](https://github.com/americanexpress/one-app-cli/commit/1cc171bffec3c4f4da8aa3689eb9c3e10dd89870))
* **devBundler:** add dev bundler ([ff0b49d](https://github.com/americanexpress/one-app-cli/commit/ff0b49dd83d0bbe02c87d111c3fbd512fa53200a))





# 1.2.0 (2023-02-16)


### Bug Fixes

* **indexloader:** regex was failing on windows machines ([#494](https://github.com/americanexpress/one-app-cli/issues/494)) ([aadd37f](https://github.com/americanexpress/one-app-cli/commit/aadd37faea92fce9b6c691c061441e1624f40cb1))


### Features

* **bundler:** enable dev bundler through flag ([1cc171b](https://github.com/americanexpress/one-app-cli/commit/1cc171bffec3c4f4da8aa3689eb9c3e10dd89870))
* **devBundler:** add dev bundler ([ff0b49d](https://github.com/americanexpress/one-app-cli/commit/ff0b49dd83d0bbe02c87d111c3fbd512fa53200a))





# 1.1.0 (2023-01-05)


### Features

* **bundler:** enable dev bundler through flag ([1cc171b](https://github.com/americanexpress/one-app-cli/commit/1cc171bffec3c4f4da8aa3689eb9c3e10dd89870))
* **devBundler:** add dev bundler ([ff0b49d](https://github.com/americanexpress/one-app-cli/commit/ff0b49dd83d0bbe02c87d111c3fbd512fa53200a))
