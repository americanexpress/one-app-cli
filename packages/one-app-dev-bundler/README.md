# One App Dev Bundler

This package provides a function to bundle modules using esBuild.

It is not intended to be installed or called directly in a module.

Instead, this package is consumed by `@americanexpress/one-app-bundler`.

See '../one-app-bundler/README.md' for details
