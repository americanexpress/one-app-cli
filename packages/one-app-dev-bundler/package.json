{
  "name": "@americanexpress/one-app-dev-bundler",
  "version": "1.5.3",
  "description": "A development bundler focussed on speed and modern features.",
  "main": "index.js",
  "bin": {
    "bundle-analysis": "bin/bundle-analysis.js"
  },
  "scripts": {
    "test": "npm run test:lint && npm run test:unit && npm run test:git-history",
    "test:lint": "eslint --ext js,jsx,mjs,snap,md .",
    "test:unit": "jest",
    "test:git-history": "commitlint --from origin/main --to HEAD"
  },
  "keywords": [
    "one",
    "amex",
    "bundler",
    "react",
    "esbuild"
  ],
  "contributors": [
    "Michael Rochester <michael.rochester@aexp.com> (https://github.com/code-forger)"
  ],
  "module": "true",
  "type": "module",
  "dependencies": {
    "@americanexpress/one-app-locale-bundler": "^6.6.0",
    "@esbuild-plugins/node-globals-polyfill": "^0.1.1",
    "@esbuild-plugins/node-modules-polyfill": "^0.2.2",
    "@fal-works/esbuild-plugin-global-externals": "^2.1.2",
    "@fullhuman/postcss-purgecss": "^3.0.0",
    "@swc/core": "^1.2.196",
    "acorn": "^8.8.0",
    "acorn-walk": "^8.2.0",
    "chalk": "^4.1.2",
    "cssnano": "^5.1.15",
    "esbuild": "^0.15.9",
    "esbuild-plugin-svgr": "^1.0.1",
    "filesize": "^9.0.1",
    "glob-all": "^3.3.0",
    "lodash.snakecase": "^4.1.1",
    "ms": "^2.1.3",
    "postcss": "^7.0.32",
    "postcss-modules": "^3.2.2",
    "read-pkg-up": "^9.0.0",
    "regenerator-runtime": "^0.13.9",
    "sass": "^1.68.0",
    "ssri": "^7.1.0",
    "ws": "^8.4.0"
  },
  "devDependencies": {
    "@testing-library/react": "^12.1.5",
    "hoist-non-react-statics": "^3.3.2",
    "immutable": "^4.0.0",
    "mock-fs": "^5.1.1",
    "prop-types": "^15.8.1",
    "react": "^16.7.0",
    "react-dom": "^16.7.0"
  },
  "license": "Apache-2.0",
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional"
    ],
    "rules": {
      "scope-case": [
        2,
        "always",
        [
          "pascal-case",
          "camel-case",
          "kebab-case"
        ]
      ],
      "body-max-line-length": [
        0
      ]
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
