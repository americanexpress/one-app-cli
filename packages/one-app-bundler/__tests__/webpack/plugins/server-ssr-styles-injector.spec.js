/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { emptyAggregatedStyles, getAggregatedStyles } from '@americanexpress/one-app-dev-bundler';
import fs from 'node:fs';
import ServerSsrStylesInjectorPlugin from '../../../webpack/plugins/server-ssr-styles-injector.js';
import {
  stylesPlaceholderUUID,
} from '../../../webpack/loaders/index-server-ssr-styles-placeholder-loader.js';

jest.mock('@americanexpress/one-app-dev-bundler', () => (
  {
    emptyAggregatedStyles: jest.fn(),
    getAggregatedStyles: jest.fn(() => JSON.stringify([{ css: '* {color: red}', digest: 'digest1Mock' }, { css: '* {color: green}', digest: 'digest2Mock' }])),
  }
));

const mockStylesPlaceholderUUID = stylesPlaceholderUUID;
jest.mock('node:fs', () => ({
  promises: {
    readFile: jest.fn(async () => `
    // replaced
    ssrStyles = '${mockStylesPlaceholderUUID}'
    
    =====
    
    // replaced
    ssrStyles = "${mockStylesPlaceholderUUID}"
    
    =====
    
    // not replaced
    ssrStyles = ${mockStylesPlaceholderUUID}
    `),
    writeFile: jest.fn(),
  },
}));

describe('ServerSsrStylesInjectorPlugin', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('should Register an assetEmitted tap hook', () => {
    const plugin = new ServerSsrStylesInjectorPlugin();

    const mockCompiler = {
      hooks: {
        assetEmitted: {
          tap: jest.fn(),
        },
      },
    };

    plugin.apply(mockCompiler);

    expect(mockCompiler.hooks.assetEmitted.tap).toHaveBeenCalledTimes(1);
    expect(mockCompiler.hooks.assetEmitted.tap.mock.calls[0][0]).toBe('ServerSsrStylesInjectorPlugin');
  });

  it('The registered function should inject the aggregated styles into the placeholder for .node.js files', async () => {
    expect.assertions(7);
    const plugin = new ServerSsrStylesInjectorPlugin();

    let registeredFn;
    const mockCompiler = {
      hooks: {
        assetEmitted: {
          tap: jest.fn((_, fn) => {
            registeredFn = fn;
          }),
        },
      },
    };

    plugin.apply(mockCompiler);

    expect(registeredFn).not.toBe(undefined);

    await registeredFn('fileNameMock', { targetPath: 'target/Path/Mock.node.js' });

    expect(fs.promises.readFile).toHaveBeenCalledTimes(1);
    expect(fs.promises.writeFile).toHaveBeenCalledTimes(1);

    expect(fs.promises.writeFile.mock.calls[0][1]).toMatchSnapshot();

    expect(mockCompiler.hooks.assetEmitted.tap.mock.calls[0][0]).toBe('ServerSsrStylesInjectorPlugin');

    expect(getAggregatedStyles).toHaveBeenCalledTimes(1);
    expect(emptyAggregatedStyles).toHaveBeenCalledTimes(1);
  });
  it('The registered function should do nothing for non .node.js files', async () => {
    expect.assertions(5);
    const plugin = new ServerSsrStylesInjectorPlugin();

    let registeredFn;
    const mockCompiler = {
      hooks: {
        assetEmitted: {
          tap: jest.fn((_, fn) => {
            registeredFn = fn;
          }),
        },
      },
    };

    plugin.apply(mockCompiler);

    expect(registeredFn).not.toBe(undefined);

    await registeredFn('fileNameMock', { targetPath: 'target/Path/Mock.note.js' });
    await registeredFn('fileNameMock.node.js', { targetPath: 'target/Path/Mock.browser.js' });
    await registeredFn('fileNameMock', { targetPath: 'target/Path/Mock.legacyBrowser.js' });
    await registeredFn('fileNameMock', { targetPath: 'target/Path/Mock.something.js' });

    expect(fs.promises.readFile).toHaveBeenCalledTimes(0);
    expect(fs.promises.writeFile).toHaveBeenCalledTimes(0);

    expect(getAggregatedStyles).toHaveBeenCalledTimes(0);
    expect(emptyAggregatedStyles).toHaveBeenCalledTimes(0);
  });
});
