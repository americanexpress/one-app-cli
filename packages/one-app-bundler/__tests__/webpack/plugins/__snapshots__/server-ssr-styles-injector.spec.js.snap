// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerSsrStylesInjectorPlugin The registered function should inject the aggregated styles into the placeholder for .node.js files 1`] = `
"
    // replaced
    ssrStyles = {
  aggregatedStyles: [{\\"css\\":\\"* {color: red}\\",\\"digest\\":\\"digest1Mock\\"},{\\"css\\":\\"* {color: green}\\",\\"digest\\":\\"digest2Mock\\"}],
  getFullSheet: function getFullSheet() {
    return this.aggregatedStyles.reduce((acc, { css }) => acc + css, '');
  },
}
    
    =====
    
    // replaced
    ssrStyles = {
  aggregatedStyles: [{\\"css\\":\\"* {color: red}\\",\\"digest\\":\\"digest1Mock\\"},{\\"css\\":\\"* {color: green}\\",\\"digest\\":\\"digest2Mock\\"}],
  getFullSheet: function getFullSheet() {
    return this.aggregatedStyles.reduce((acc, { css }) => acc + css, '');
  },
}
    
    =====
    
    // not replaced
    ssrStyles = 2574b873-59dc-4da6-8692-d767a2484dac
    "
`;
